src/game_logic/contracts.ts

```diff
@@ -0,0 +1,112 @@
+/**
+ * This module defines common functions for contracts
+ */
+
+import { GameState, ProjectType, Tile, TileType } from '.'
+import { tileCatalog } from '../data/tile-catalog'
+import { mapIncr } from './shared'
+
+type Comparator = (actual: number, desired: number) => boolean
+export const COMP_GTE: Comparator = (x, y) => x >= y
+export const COMP_EQ: Comparator = (x, y) => x === y
+
+function countTilesByTags (tiles: readonly Tile[]): Map<string, number> {
+  const result = new Map()
+  for (const tile of tiles) {
+    for (const tag of tileCatalog[tile.type].tags) {
+      mapIncr(result, tag)
+    }
+  }
+  return result
+}
+
+function countProjectsCompletedByType (projectsComplete: readonly ProjectType[]): Map<ProjectType, number> {
+  const result = new Map<ProjectType, number>()
+  for (const project of projectsComplete) {
+    mapIncr(result, project)
+  }
+  return result
+}
+
+function countTilesByType (tiles: readonly Tile[]): Map<TileType, number> {
+  const result = new Map()
+  for (const tile of tiles) {
+    mapIncr(result, tile.type)
+  }
+  return result
+}
+
+function compareCounts<K extends string> (desiredCounts: Record<K, number>, actualCounts: Map<K, number>, comparator: Comparator): boolean {
+  let satisfied = true
+  for (const key in desiredCounts) {
+    const desired = desiredCounts[key]
+    const actual = actualCounts.get(key)
+    if (actual) {
+      satisfied = satisfied && comparator(actual, desired)
+    } else {
+      satisfied = false
+      break
+    }
+  }
+  return satisfied
+}
+
+type DesiredCounts = Record<string, number>
```

Comment:
`string` is a little bleh, but even if we used WhateverType here TS wouldn't differentiate between them. Sad